### ISSUE TITLE
"S3 Bucket Should Have Bucket Policy" Does Not Handle `!Ref` Properly

### ISSUE THREAD

**Issue by chrisoverzero** [Checkmarx/kics#4303]
### Expected Behavior

Bucket policies which refer to their buckets with `!Ref` are found by query "37fa8188-738b-42c8-bf82-6334ea567738".

### Actual Behavior

Buckets with policies are reported as having no policy.

### Steps to Reproduce the Problem

```yaml
AWSTemplateFormatVersion: 2010-09-09

Resources:
  EncryptionKey:
    Type: AWS::KMS::Key
    Properties:
      Description: An encryption key for this example.
      EnableKeyRotation: true
      KeyPolicy:
        Version: 2012-10-17
        Id: EncryptionKeyPolicy
        Statement:
          Effect: Allow
          Action: kms:*
          Principal:
            AWS: !Ref AWS::AccountId
          Resource: '*'
  Bucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketEncryption:
        ServerSideEncryptionConfiguration:
        - ServerSideEncryptionByDefault:
            SSEAlgorithm: aws:kms
            KMSMasterKeyID: !Ref EncryptionKey
      LoggingConfiguration:
        DestinationBucketName: !ImportValue S3AccessLogsBucket
      VersioningConfiguration:
        Status: Enabled
      PublicAccessBlockConfiguration:
        BlockPublicAcls: true
        BlockPublicPolicy: true
        IgnorePublicAcls: true
        RestrictPublicBuckets: true
  BucketPolicy:
    Type: AWS::S3::BucketPolicy
    Properties:
      Bucket: !Ref Bucket
      PolicyDocument:
        Statement:
          Action: s3:*
          Effect: Deny
          Principal: '*'
          Resource:
          - !GetAtt Bucket.Arn
          - !Sub ${Bucket.Arn}/*
          Condition:
            Bool:
              aws:SecureTransport: 'false'
```

This small-but-not-quite-minimized template fails "S3 Bucket Should Have Bucket Policy" and (maybe as a consequence?) "S3 Bucket Without SSL In Write Actions". This may be related to #3760.

```json
[
  {
    "query_name": "S3 Bucket Without SSL In Write Actions",
    "query_id": "38c64e76-c71e-4d92-a337-60174d1de1c9",
    "query_url": "https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html",
    "severity": "HIGH",
    "platform": "CloudFormation",
    "category": "Encryption",
    "description": "S3 Buckets should enforce encryption of data transfers using Secure Sockets Layer (SSL)",
    "description_id": "907676d7",
    "cis_description_id": "",
    "cis_description_title": "",
    "cis_description_text": "",
    "files": "<omitted>"
  },
  {
    "query_name": "S3 Bucket Should Have Bucket Policy",
    "query_id": "37fa8188-738b-42c8-bf82-6334ea567738",
    "query_url": "https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html",
    "severity": "MEDIUM",
    "platform": "CloudFormation",
    "category": "Insecure Defaults",
    "description": "Checks if S3 Bucket has the same name as a Bucket Policy, if it has, S3 Bucket has a Bucket Policy associated",
    "description_id": "2b429546",
    "cis_description_id": "",
    "cis_description_title": "",
    "cis_description_text": "",
    "files": "<omitted>"
  }
]
```

This may be due to an unused variable in the specification:

```go
bucketName(bucketName) {
	resource := input.document[i].Resources[name]
	resource.Type == "AWS::S3::BucketPolicy"
	bucket := resource.Properties.Bucket
	contains(bucket, "!Ref")
	bucketN := replace(bucket, "!Ref ", "")
	bucket == bucketName
}
```

See how `bucketN` is created but not used in the comparison? I don't think there's a unit test for the case of `!Ref`.

### Specifications

-   Version: Keeping Infrastructure as Code Secure 1.4.3
-   Platform: CloudFormation
-   Subsystem: Subsystem: macOS Catalina (10.15.7), darwin_x64 (Maybe? I don't know what Subsystem is.)

**Comment by rogeriopeixotocx** [comment]
@chrisoverzero thank you for reporting this, we'll take a look as soon as possible

**Comment by rafaela-soares** [comment]
Hi, @chrisoverzero 🙂 Thank you so much for reaching us. I hope you are doing well.

This sample returns a result for query "S3 Bucket Should Have Bucket Policy" because there is no "BucketName" defined and this query searches for it. It fails in the following rule:

```
CxPolicy[result] {
	resource := input.document[i].Resources[name]
	resource.Type == "AWS::S3::Bucket"
	not common_lib.valid_key(resource.Properties, "BucketName")

	result := {
		"documentId": input.document[i].id,
		"searchKey": sprintf("Resources.%s.Properties", [name]),
		"issueType": "MissingAttribute",
		"keyExpectedValue": sprintf("'Resources.%s.Properties.BucketName' is defined", [name]),
		"keyActualValue": sprintf("'Resources.%s.Properties.BucketName' is not defined", [name]),
	}
}

```

We will take a deep look to check the need for the BucketName definition. Additionally, we will verify the unused variable `bucketN`.



Regarding "S3 Bucket Without SSL In Write Actions", KICS was not checking cases where a Statement is an object (which occurs in your sample). We will correct it.

Thank you so much for catch and report both bugs.
